### ISSUE TITLE
Add a specific OG image and description for project list pages

### ISSUE THREAD

**Issue by michaelrambeau** [bestofjs/bestofjs#210]
Thanks to the dynamic generation of images provided by [Satori project](https://og-playground.vercel.app/), we provide a specific OG (Open Graph) image for the 2 following pages:

- the top page: `/` path
- the project's detail page `/projects/:slug` path
 
It would be great to get a specific OG image for project list pages (`/projects` path, with or without query string), showing an overview of the 3 first projects of the list.

Right now, it's the top page OG image that is displayed, because metadata is inherited from the parent page and `/app/projects/page.tsx` does not specify any

Docs about metadata: https://nextjs.org/docs/app/api-reference/file-conventions/metadata/opengraph-image

 ### Screenshots

Top page

Image URL: https://bestofjs.org/api/og

<img width="629" alt="image" src="https://github.com/bestofjs/bestofjs/assets/5546996/f29cf81a-2fa9-411f-8cf6-bf7b76ed0000">

For React project

Image URL: https://bestofjs.org/api/og/projects/react

<img width="421" alt="image" src="https://github.com/bestofjs/bestofjs/assets/5546996/765be5d8-9d01-47f3-ad15-72fb90b38433">

### To be defined

What to show as the caption, before the 3 rows?
Something similar to what we show in the `/projects` page URLs?

There are different scenarios:

- no query string at all: it's the list of all available projects
- one or several tags
- without or without `query` parameter when showing search results
 
For example, for `Vue.js` projects, the tag icon, followed by the name of the tag:

![image](https://github.com/bestofjs/bestofjs/assets/5546996/7b5202cb-1374-4b48-bb66-a00342dd1718)

**Comment by Evomatic** [comment]
Hello @michaelrambeau 

After taking a look, I want to make sure I have a correct understanding of what we want to do here!

So we want to add a new OG image that is specific for the path `/projects`, because right now, linking `https://bestofjs.org/projects` gives us the same OG image used at the top page. 

Things to consider for this new specific OG image:

- If no query string, list all available projects. Or do we want to list the first 3 (depending on sort method), similar to the top page list?
- If tags are included (one to three tags) how should we show that in the caption?
- If `query` parameter is applicable, how should we show this?

 As you mentioned I think doing something similar to what is shown for `/projects` is good. For example, if using tags then show tag icon, followed by name of tag. Then show the list projects with the tag below.

If no tags selected then show "All Projects" followed by number of projects

For `query` search results, unless I am missing something, it seems I can only search via tags, and if I search for a project by name I am brought straight to the project page. Either way, the scenario seems similar to choosing tags and what we see when linking `/projects/:slug`

**Comment by Evomatic** [comment]
Looking at the code, i see the file structure:

```
og
├── og-image-layout.tsx
├── og-utils.tsx
├── projects
        └── [slug]
                └── route.tsx
└── route.tsx
```

Creating a new OG image, do we need to create a new endpoint and add it to the above directory?

**Comment by michaelrambeau** [comment]
It's possible to search for projects by text:

<img width="882" alt="image" src="https://github.com/bestofjs/bestofjs/assets/5546996/c1a364da-5e66-43ed-bc3f-17e3872db8af">

And it generates an a `/projects` URL with a query parameter:

https://bestofjs.org/projects?query=jest

The query string parameters can also include the `tags` and a sort option (by default, projects are sorted by number of stars but users can change it):

<img width="510" alt="image" src="https://github.com/bestofjs/bestofjs/assets/5546996/22137a71-26ed-42ac-94cd-8e61851421f7">

About the file system, the image must be available at `/api/og/projects?tags=...&query=...&sort=...`.

So we need a `route.tsx` file under `projects` (basically we mirror what we have in `app` folder using `route.tsx` files instead of `pages.tsx`)

**Comment by michaelrambeau** [comment]
For the text to be displayed before the list of projects, we can use something similar to:

https://github.com/bestofjs/bestofjs/blob/289bb7e6213d5f9ae922e3d3d9d446de4aa9bbff/apps/bestofjs-nextjs/src/app/projects/page.tsx#L68-L77

**Comment by Evomatic** [comment]
@michaelrambeau

Ok, I see how to do it now, thanks!

> So we need a route.tsx file under projects (basically we mirror what we have in app folder using route.tsx files instead of pages.tsx)

Alright, so we want to add `route.tsx` to `/app/api/og/projects`, (which will create an endpoint for the new og image at `/projects` and in this file we can define the logic for image generation and how it should behave for the different scenarios?

**Comment by michaelrambeau** [comment]
> @michaelrambeau
> 
> Ok, I see how to do it now, thanks!
> 
> > So we need a route.tsx file under projects (basically we mirror what we have in app folder using route.tsx files instead of pages.tsx)
> 
> Alright, so we want to add `route.tsx` to `/app/api/og/projects`, (which will create an endpoint for the new og image at `/projects` and in this file we can define the logic for image generation and how it should behave for the different scenarios? (similar to how it is done for the other `route.tsx` files)

Yes, the end-point with grab the URL search parameters and call the projects API with exactly the same request as the page does, except the results would be limited to 3 results (the projects we want to display in the card).

That means we may need to share some logic to parse the search parameters and fetch the projects between `page.tsx` and `route.tsx` (it can be done in the next iteration, do it simple at first)

**Comment by Evomatic** [comment]
After experimenting a bit, I have a simple OG image rendering for `/projects`!
![image](https://github.com/bestofjs/bestofjs/assets/27636896/383d3173-5e28-4b16-bf1c-2b4b4f02dafc)

> That means we may need to share some logic to parse the search parameters and fetch the projects between page.tsx and route.tsx (it can be done in the next iteration, do it simple at first)

For getting the search params I was able to do so with the JS URL interface:
```
const searchParams = new URLSearchParams(_req.url)
``` 
or 
```
 const { searchParams } = new URL(_req.url)
```
But I am not sure if this is the best way to do so. In [page.tsx](https://github.com/bestofjs/bestofjs/blob/develop/apps/bestofjs-nextjs/src/app/projects/page.tsx) there is a a `parseSearchParams` method. I thought I could import and use it but there is a 'Types' error:
```Argument of type 'URLSearchParams' is not assignable to parameter of type 'ProjectPageSearchParams'.
  Types of property 'sort' are incompatible.
    Type '() => void' is not assignable to type 'string'.ts(2345)
```

In this case, is sharing the `parseSearchParams` method something we want to do? Or is it preferred to use similar logic to create a method specifically for parsing params coming from JS URL Interface?

**Comment by michaelrambeau** [comment]
My first idea would be, given that we have this type:

```ts
// Raw params values from URLSearchParams.get() and .getAll() methods, before any parsing
export type ProjectPageSearchParams = {
  tags?: string[];
  query?: string;
  page?: string;
  limit?: string;
  sort?: string;
};
```

Create a `ProjectPageSearchParams` object, reading all properties`tags`, `query`... one by one, from the `searchParams` you got, using `searchParams.get() and `searchParams.getAll()` (for tags)

And then you could call the shared function `parseSeachParams`?

We could see later if we could clean the redundancies in the code.

**Comment by michaelrambeau** [comment]
Fixed by #215 and #232
